shell/browser/api/atom_api_service_worker_context.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2019 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/atom_api_service_worker_context.h"
+
+#include <string>
+#include <utility>
+
+#include "chrome/browser/browser_process.h"
+#include "content/public/browser/console_message.h"
+#include "content/public/browser/storage_partition.h"
+#include "native_mate/converter.h"
+#include "native_mate/dictionary.h"
+#include "native_mate/handle.h"
+#include "shell/browser/atom_browser_context.h"
+#include "shell/common/gin_converters/value_converter.h"
+#include "shell/common/gin_helper/object_template_builder.h"
+#include "shell/common/node_includes.h"
+
+namespace electron {
+
+namespace api {
+
+namespace {
+
+std::string MessageSourceToString(
+    const blink::mojom::ConsoleMessageSource source) {
+  if (source == blink::mojom::ConsoleMessageSource::kXml)
+    return "xml";
+  if (source == blink::mojom::ConsoleMessageSource::kJavaScript)
+    return "javascript";
+  if (source == blink::mojom::ConsoleMessageSource::kNetwork)
+    return "network";
+  if (source == blink::mojom::ConsoleMessageSource::kConsoleApi)
+    return "console-api";
+  if (source == blink::mojom::ConsoleMessageSource::kStorage)
+    return "storage";
+  if (source == blink::mojom::ConsoleMessageSource::kAppCache)
+    return "app-cache";
+  if (source == blink::mojom::ConsoleMessageSource::kRendering)
+    return "rendering";
+  if (source == blink::mojom::ConsoleMessageSource::kSecurity)
+    return "security";
+  if (source == blink::mojom::ConsoleMessageSource::kDeprecation)
+    return "deprecation";
+  if (source == blink::mojom::ConsoleMessageSource::kWorker)
+    return "worker";
+  if (source == blink::mojom::ConsoleMessageSource::kViolation)
+    return "violation";
+  if (source == blink::mojom::ConsoleMessageSource::kIntervention)
+    return "intervention";
+  if (source == blink::mojom::ConsoleMessageSource::kRecommendation)
+    return "recommendation";
+  return "other";
+}
+
+base::DictionaryValue ServiceWorkerRunningInfoToDict(
+    const content::ServiceWorkerRunningInfo& info) {
+  base::DictionaryValue dict;
+  dict.SetStringPath("scriptUrl", info.script_url.spec());
+  dict.SetStringPath("scope", info.scope.spec());
+  dict.SetIntPath("renderProcessId", info.render_process_id);
+  return dict;
+}
+
+}  // namespace
+
+ServiceWorkerContext::ServiceWorkerContext(v8::Isolate* isolate,
+                                           AtomBrowserContext* browser_context)
+    : browser_context_(browser_context), weak_ptr_factory_(this) {
+  Init(isolate);
+  service_worker_context_ =
+      content::BrowserContext::GetDefaultStoragePartition(browser_context_)
+          ->GetServiceWorkerContext();
```

Comment:
I don't think this object is guaranteed to have a shorter lifetime than the `ServiceWorkerContext` to which it refers, since it's retained by V8.